packages/@ember/-internals/runtime/lib/mixins/array.ts

```diff
@@ -1775,41 +1929,45 @@ let NativeArray = Mixin.create(MutableArray, Observable, {
 // Remove any methods implemented natively so we don't override them
 const ignore = ['length'];
 NativeArray.keys().forEach((methodName) => {
-  if (Array.prototype[methodName]) {
+  // SAFETY: It's safe to read unknown properties from an object
+  if ((Array.prototype as any)[methodName]) {
     ignore.push(methodName);
   }
 });
 
 NativeArray = NativeArray.without(...ignore);
 
-let A;
+let A: <T>(arr?: Array<T>) => NativeArray<T>;
 
 if (ENV.EXTEND_PROTOTYPES.Array) {
   NativeArray.apply(Array.prototype, true);
 
-  A = function (arr) {
+  A = function <T>(this: unknown, arr?: Array<T>) {
     assert(
       'You cannot create an Ember Array with `new A()`, please update to calling A as a function: `A()`',
       !(this instanceof A)
     );
 
-    return arr || [];
+    // SAFTEY: Since we are extending prototypes all true native arrays are Ember NativeArrays
+    return (arr || []) as NativeArray<T>;
   };
 } else {
-  A = function (arr) {
+  A = function <T>(this: unknown, arr?: Array<T>) {
     assert(
       'You cannot create an Ember Array with `new A()`, please update to calling A as a function: `A()`',
       !(this instanceof A)
     );
 
-    if (!arr) {
-      arr = [];
+    if (isEmberArray(arr)) {
```

Comment:
Ideally we'd be more consistent in our use of this method vs EmberArray.detect.

Reply:
Ideally we will soon burn all of this out of existence. 😏

Reply:
Sadly this breaks EmberData's migration path :( 

Data Issue: https://github.com/emberjs/data/issues/8202
PR to restore: https://github.com/emberjs/ember.js/pull/20219